File: Toy_Example/wsOFMM_main_latent.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Weighted Supervised OFMM Main %
% Programmer: SW                %
% Data: Simulations             %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% wsOFMM_main_latent takes in scenario, sim_n, and samp_n as command line 
% array index arguments. It reads in the simulated dataset corresponding to 
% the given array and scenario, runs the wsOFMM model with the latent 
% variables, and saves the MCMC output and posterior results as files.
% Inputs:
%   scenario: Command line argument indicating weighting scenario
%   sim_n: Command line argument indicating simulation array index
%   samp: Command line argument indicating sample index
% Outputs: saves MCMC output in a file named 'wsOFMM_MCMC...' and saves 
% posterior results in a file named 'wsOFMM_results...'
function wsOFMM_main_latent(scenario, sim_n, samp_n)
    % set seed
    rng(sim_n, 'twister');

    %% Load simulated data and check if file already exists
    % Input directory
    in_dir = "/n/holyscratch01/stephenson_lab/Users/stephwu18/wsOFMM/Toy_Example/";   
    % Output directory 
    out_dir = "/n/holyscratch01/stephenson_lab/Users/stephwu18/wsOFMM/Toy_Example/";  
    if samp_n > 0   % If working with sample 
        samp_data = importdata(strcat(in_dir, 'simdata_scen', num2str(scenario), '_iter', num2str(sim_n), '_samp', num2str(samp_n), '.mat'));
        already_done = isfile(strcat(out_dir, 'wsOFMM_latent_results_scen', num2str(scenario), '_iter', num2str(sim_n), '_samp', num2str(samp_n), '.mat'));
    else            % If working with population
        samp_data = importdata(strcat(in_dir, 'simdata_scen', num2str(scenario), '_iter', num2str(sim_n), '.mat'));
        already_done = isfile(strcat(out_dir, 'wsOFMM_latent_results_scen', num2str(scenario), '_iter', num2str(sim_n), '.mat'));
    end
    
    % Run the model if the results file does not already exist
    if already_done
        % If the results file already exists, print out statement
        disp(strcat('Scenario ', num2str(scenario), ' iter ', num2str(sim_n), ' samp ', num2str(samp_n), ' already exists.'));
    else
        %% Get data variables
        k_max = 50;    % Upper limit for number of classes
        data_vars = wtd_get_data_vars_latent(samp_data);

        %% Initialize priors and variables for OFMM model
        sp_k = k_max;                   % Denom constant to restrict alpha size and num classes for OFMM model
        alpha = ones(1, k_max) / sp_k;  % Hyperparam for class membership probs. R package 'Zmix' allows diff options
        eta = ones(1, data_vars.d_max); % Hyperparam for item response probs. 
        OFMM_params = wtd_init_OFMM_params_latent(data_vars, k_max, alpha, eta);

        %% Initialize priors and variables for probit model
        q_dem = dummyvar(samp_data.true_Si);        % Matrix of demographic covariates in cell-means format. Default contains subpop.                               
        clear samp_data;                            % Reduce memory burden
        S = size(q_dem, 2);                         % Number of demographic covariates in the probit model
        p_cov = k_max + S;                          % Number of covariates in probit model
        mu_0 = normrnd(0, 1, [p_cov, 1]);           % Mean hyperparam drawn from MVN(0,1)
        Sig_0 = 1 ./ gamrnd(5/2, 2/5, [p_cov, 1]);  % Var hyperparam drawn from MVGamma(shape=5/2, scale=5/2)
        Sig_0 = diag(Sig_0);                        % Assume indep components. (pcov)x(pcov) matrix of variances. 
        probit_params = init_probit_params_latent(data_vars, k_max, q_dem, mu_0, Sig_0, OFMM_params);

        %% Run adaptive sampler to obtain number of classes
        n_runs = 25000;  % Number of MCMC iterations
        burn = 15000;    % Burn-in period
        thin = 5;        % Thinning factor
        [MCMC_out, ~, ~] = wtd_run_MCMC_latent(data_vars, OFMM_params, probit_params, n_runs, burn, thin, k_max, q_dem, p_cov, alpha, eta, mu_0, Sig_0);
        k_fixed = round(median(sum(MCMC_out.pi > 0.05, 2))); % Obtain fixed number of classes to use in the fixed sampler
        clear OFMM_params probit_params MCMC_out;           % Reduce memory burden

        %% Run fixed sampler to obtain posteriors and save output
        % Initialize OFMM model using fixed number of classes  
        sp_k = k_fixed;                   % Denom constant to restrict alpha size and num classes for OFMM model
        alpha = ones(1, k_fixed) / sp_k;  % Hyperparam for class membership probs. R package 'Zmix' allows diff options
        OFMM_params = wtd_init_OFMM_params_latent(data_vars, k_fixed, alpha, eta);

        % Initialize probit model using fixed number of classes
        p_cov = k_fixed + S;                        % Number of covariates in probit model
        mu_0 = normrnd(0, 1, [p_cov, 1]);           % Mean hyperparam drawn from MVN(0,1)
        Sig_0 = 1 ./ gamrnd(5/2, 2/5, [p_cov, 1]);  % Var hyperparam drawn from MVInvGamma(shape=5/2, scale=5/2)
        Sig_0 = diag(Sig_0);                        % Assume indep components. (pcov)x(pcov) matrix of variances. 
        probit_params = init_probit_params_latent(data_vars, k_fixed, q_dem, mu_0, Sig_0, OFMM_params);

        % Run MCMC algorithm using fixed number of classes
        [MCMC_out, ~, ~] = wtd_run_MCMC_latent(data_vars, OFMM_params, probit_params, n_runs, burn, thin, k_fixed, q_dem, p_cov, alpha, eta, mu_0, Sig_0);
        % Save MCMC output
        if samp_n > 0  % If working with sample 
            save(strcat(out_dir, 'wsOFMM_latent_MCMC_scen', num2str(scenario), '_iter', num2str(sim_n), '_samp', num2str(samp_n)), 'MCMC_out');
        else
            save(strcat(out_dir, 'wsOFMM_latent_MCMC_scen', num2str(scenario), '_iter', num2str(sim_n)), 'MCMC_out');
        end    

        %% Post-processing to recalibrate labels and remove extraneous empty classes
        post_MCMC_out = post_process_latent(MCMC_out, data_vars, S);
        clear OFMM_params probit_params;  % Reduce memory burden

        %% Obtain posterior estimates, reduce number of classes, analyze results, and save output
        analysis = analyze_results_latent(MCMC_out, post_MCMC_out, data_vars, q_dem, S, p_cov);
        % Save parameter estimates and analysis results
        if samp_n > 0  % If working with sample 
            save(strcat(out_dir, 'wsOFMM_latent_results_scen', num2str(scenario), '_iter', num2str(sim_n), '_samp', num2str(samp_n)), 'post_MCMC_out', 'analysis');
        else
            save(strcat(out_dir, 'wsOFMM_latent_results_scen', num2str(scenario), '_iter', num2str(sim_n)), 'post_MCMC_out', 'analysis');
        end 
   end  
end


%% Miscellaneous additional code
%     % Create and save figures
%     figure; %check mixing of pi parameter
%     plot(MCMC_out.pi)
%     saveas(gcf,'wsOFMM_pis.png')
%     
%     figure; %save dendrogram of hierarchical clustering
%     dendrogram(post_MCMC_out.tree);
%     saveas(gcf,'wsOFMM_dendrogram.png')
%
%     % Compute the MSE comparing the estimated probit model to the true model 
%     y_mse = immse(analysis.Phi_med, samp_data.true_Phi)  % Phi_true must be pulled from simulated dataset
% 
%     % Testing code
%     in_dir = "C:/Users/Lang/Documents/Harvard/Research/Briana/supRPC/wsOFMM/Data/";
%     out_dir = "C:/Users/Lang/Documents/Harvard/Research/Briana/supRPC/wsOFMM/Results/";
%     n_runs = 50;  % Number of MCMC iterations
%     burn = 30;    % Burn-in period
%     thin = 2;        % Thinning factor
%     sim_n = 1;    % Simulation iteration
    

File: Toy_Example/wsOFMM_main_latent.R
#====================================================
# Weighted Supervised Overfitted Finite Mixture Model
# Stephanie Wu
# Revised 06/29/2022
#====================================================

library(R.matlab)
library(extraDistr)
library(MASS)

#=========================
# Example parameter setup
scenario <- 6
sim_n <- 1
samp_n <- 2
in_dir <- paste0(getwd(), "/")
out_dir <- in_dir

k_max <- 50
sp_k <- k_max
alpha <- rep(1/sp_k, k_max)
eta <- rep(1, 4)

mu_0 <- rnorm(n=4, mean=0, sd=1)
Sig_0 <- 1 / rgamma(n=4, shape=5/2, rate=2/5)
Sig_0 <- diag(Sig_0)


#=================================
# Set seed and load simulated data
set.seed(sim_n)
samp_data <- readMat(paste0(in_dir, "simdata_scen", scenario, 
                              "_iter", sim_n, 
                              "_samp", samp_n, ".mat"))

#===================
# Get data variables
data_vars <- wtd_get_data_vars_latent(samp_data)

#===============================================
# Initialize priors and variables for OFMM model
OFMM_params <- wtd_init_OFMM_params_latent(data_vars, k_max, alpha, eta)

#=================================================
# Initialize priors and variables for probit model
probit_params <- init_probit_params_latent(data_vars, k_max, q_dem, 
                                           mu_0, Sig_0, OFMM_params)


# wtd_get_data_vars_latent takes in sample data and outputs relevant variables for
# the wsOFMM model with the latent variable formulation
# Inputs: samp_data structural array with at least the following columns:
#   X_data: food intake data as a matrix
#   Y_data: outcome data as a vector
#   sample_wt: survey weights as a vector
# Outputs: data_vars structural array with the following fields:
#   food: matrix of food intake data
#   n: number of individuals
#   p: number of food items
#   d_max: max number of consumption levels over all items
#   d: vector of max number of levels for each food item; px1
#   y: vector of outcomes; nx1
#   wt: vector of survey weights; nx1
#   wt_kappa: vector of normalized weights; nx1
#   wt_kappa_mat: matrix of normalized weights, replicated across items; nxp
#   lin_idx: vector of linear indices for unique item-response combos; (n*p)x1
wtd_get_data_vars_latent <- function(samp_data) {
  food <- samp_data$X_data
  n <- nrow(food)
  p <- ncol(food)
  d_max <- max(c(food))
  d <- apply(food, 1, max)
  y <- samp_data$Y_data
  wt <- samp_data$sample_wt
  
  kappa <- sum(wt) / n
  wt_kappa <- wt / kappa
  wt_kappa_mat <- matrix(rep(wt_kappa, times=p), nrow=n, byrow=FALSE)
  
  data_vars <- list("food" = food, "n" = n, "p" = p, "d_max" = d_max, "d" = d,
                    "y" = y, "wt" = wt, "kappa" = kappa, "wt_kappa" = wt_kappa, 
                    "wt_kappa_mat" = wt_kappa_mat)
}


wtd_init_OFMM_params_latent <- function(data_vars, k_max, alpha, eta) {
  pi <- rdirichlet(1, alpha)
  
  c_i <- rcat(n, prob=pi)  # Draw from a Categ(pi) random variable n times
  n_ci <- numeric(k_max)
  for (k in 1:k_max) {
    n_ci[k] <- sum(data_vars$wt_kappa[c_i == k])
  }
  
  theta <- array(0, dim=c(data_vars$p, k_max, data_vars$d_max), 
                 dimnames=c("p","k_max", "d_max"))
  for (k in 1:k_max) {
    for (j in 1:data_vars$p) {
      d_j <- data_vars$d[j]
      theta[j, k, 1:d_j] <- rdirichlet(1, eta[1:d_j])
    }
  }
  
}











